log/zap/field.go

```diff
@@ -59,3 +61,158 @@ func (t trace) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	})
+
+	enc.AddArray("baggage", baggage)
+}
+
+type baggageKV struct {
+	key   string
+	value string
+}
+
+func (b baggageKV) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddString("key", b.key)
+	enc.AddString("value", b.value)
+	return nil
+}
+
+type baggageKVs []baggageKV
+
+func (b baggageKVs) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	for _, kv := range b {
+		enc.AppendObject(kv)
+	}
+	return nil
+}
+
+type logRecord opentracing.LogRecord
+
+func (l logRecord) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddTime("ts", l.Timestamp)
+	enc.AddArray("fields", logFields(l.Fields))
+	return nil
+}
+
+type logRecords []opentracing.LogRecord
+
+func (l logRecords) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	for _, record := range l {
+		enc.AppendObject(logRecord(record))
+	}
+	return nil
+}
+
+type logField struct {
+	log.Field
+}
+
+func (l logField) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddString("key", l.Key())
+	enc.AddString("value", fmt.Sprint(l.Value()))
+	return nil
+}
+
+type logFields []log.Field
+
+func (l logFields) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	for _, field := range l {
+		enc.AppendObject(logField{field})
+	}
+	return nil
+}
+
+type tags opentracing.Tags
+
+func (t tags) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	for k, v := range t {
+		enc.AddString("key", k)
+		enc.AddReflected("value", v)
+	}
+	return nil
+}
+
+type reference opentracing.SpanReference
+
+func (r reference) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	if r.Type == opentracing.ChildOfRef {
+		enc.AddString("type", "child_of")
+	} else if r.Type == opentracing.FollowsFromRef {
+		enc.AddString("type", "follows_from")
+	} else {
+		enc.AddString("type", "unknown")
+	}
+
+	if jCtx, ok := r.ReferencedContext.(jaeger.SpanContext); ok {
+		enc.AddObject("context", spanContext{spanContext: jCtx})
+	}
+
+	return nil
+}
+
+type references []opentracing.SpanReference
+
+func (r references) MarshalLogArray(enc zapcore.ArrayEncoder) error {
+	for _, spanReference := range r {
+		enc.AppendObject(reference(spanReference))
+	}
+	return nil
+}
+
+type span struct {
+	span opentracing.Span
+}
+
+// Span creates a zap.Field that marshals all information contained in a jaeger span
+func Span(s opentracing.Span) zapcore.Field {
+	if s == nil {
+		return zap.Skip()
+	}
+	return zap.Object("span", span{span: s})
+}
+
+func (s span) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	if span, ok := s.span.(*jaeger.Span); ok {
+		enc.AddObject("context", spanContext{spanContext: span.SpanContext()})
+		enc.AddString("operation_name", span.OperationName())
+		enc.AddDuration("duration", span.Duration())
+		enc.AddTime("start_time", span.StartTime())
+
+		enc.AddArray("logs", logRecords(span.Logs()))
+		enc.AddObject("tags", tags(span.Tags()))
+		enc.AddArray("span_refs", references(span.References()))
+	} else {
+		enc.AddString("err", "Non Jaeger Span")
+	}
+	return nil
+}
```

Comment:
I am not convinced this deep integration with zap.Field is necessary / useful. Couldn't we just serialize the span to JSON and write that?

Reply:
It will be slower - do we already have a means of deserializing cleanly to JSON? (with proper handling of trace bytes/human readable timestamps/etc ?)